keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1046 @@
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+#### Story 1
+
+I'm an application developer building a Kubernetes controller that I want to deploy into my cluster.  I want to make sure that my controller securely authenticates to kube-apiserver using mTLS.
+
+I add a new projected volume to my pod, containing the information that client-go expects at a well-known path.  I set `noAutomountServiceAccountToken` because I want to ensure that my pod cannot use bearer tokens for authentication.
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: pod-certificates-example
+spec:
+  restartPolicy: OnFailure
+  automountServiceAccountToken: false
+  containers:
+  - name: main
+    image: debian
+    command: ['sleep', 'infinity']
+    volumeMounts:
+    - name: kube-apiserver-client-certificate
+      mountPath: /var/run/secrets/kubernetes.io/api-mtls-access
+  volumes:
+  - name: kube-apiserver-client-certificate
+    projected:
+      sources:
+      - podCertificate:
+          signerName: "kubernetes.io/kube-apiserver-client-pod"
+          credentialBundlePath: client-credentials.pem
+      - configMap:
+          localObjectReference: kube-root-ca.crt
+          items:
+          - key: ca.crt
+            path: kube-apiserver-root-certificate.pem
+      - downwardAPI:
+          items:
+          - path: namespace
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+```
+
+I ensure that my pod uses client-go's InClusterConfig, and client-go automatically uses the mTLS client certificate to authenticate to kube-apiserver.
+
+### Risks and Mitigations
+
+**Scalability**: In the current design and draft implementation, Kubelet holds
+the private keys for each pod on the node in memory.  If Kubelet restarts, then
+the private keys are lost, and Kubelet will need to immediately generate new
+keys and request new certificates for each volume on the pod.  If Kubelet
+crashloops, this can generate high CertificateSigningRequest traffic on
+kube-apiserver.  This can potentially be mitigated by giving Kubelet the
+capability to read back the existing key from the projected volume on startup.
+
+## Design Details
+
+### X.509 PodIdentity Extension
+
+The CNCF has an existing ASN.1 Private Enterprise Number registered for use by
+the Kubernetes project (`1.3.6.1.4.1.57683`).  In order to cleanly embed pod
+identity information into an X.509 certificate or certificate request, we can
+define a new ASN.1 OID (currently `1.3.6.1.4.1.57683.1`, though we need to
+define a management policy for the PEN) and specify that it contains the
+following DER-encoded structure:
+
+```go
+type podIdentityASN1 struct {
+	Namespace          string `asn1:"utf8"`
+	ServiceAccountName string `asn1:"utf8"`
+	PodName            string `asn1:"utf8"`
+	PodUID             string `asn1:"utf8"`
+	NodeName           string `asn1:"utf8"`
```

Comment:
We included Node UID in https://github.com/kubernetes/kubernetes/pull/120780/files#diff-0c530c4dd8479db341cd002079be7ab62e34d49b42540785df9ad17431c8f108R243 and should probably do the same here.

Reply:
Are all these fields required? If they are, I think we might one day regret not making the node name / uid optional.

Reply:
In the extension data itself, all fields are optional, in that you can always leave them as the empty string.  (I don't think the ASN.1 parser in Go tries to handle field presence or absence).

If the extension appears in a CertificateSigningRequest from a kubelet, the node restriction plugin will enforce that all the fields are populated, and consistent with reality.

If the extension appears in an issued certificate, I think it's up to the component that's validating the certificate to decide which fields it cares about.  For the specific example of certificates issued by the `kubernetes.io/kube-apiserver-client-pod` signer/approver, we probably want kube-apiserver to check all of the fields before letting the request past the authentication stage.

Reply:
I think clearly documenting a minimum, non-negotiable subset of fields that are provably sufficient to guarantee unambiguous pod<->cert binding in Kubernetes is required, or including this ASN.1 OID field as an overall KEP requirement isn't much use.

Reply:
Sure, I can add a detailed proposal for exactly which fields need to be set in order for kube-apiserver to allow authentication via a certificate containing a PodIdentity extension.

Reply:
Based on discussion in API review, we are going to go with a dedicated PodCertificateRequest type, rather than encoding these details in the existing CertificateSigningRequest spec.request field.  This drastically lowers the importance of the PodIdentity extension, since it will only be used by the kube-apiserver-client-pod signer, not every third-party signer that wants to integrate with pod certificates.